src/libraries/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs

```diff
@@ -149,79 +151,82 @@ internal void InitAssemblyMethods(XmlMapping[] xmlMappings)
+                    if (serializer == null)
+                    {
+                        if (XmlSerializer.Mode == SerializationMode.PreGenOnly)
+                        {
+                            throw new Exception(SR.Format(SR.FailLoadAssemblyUnderPregenMode, serializerName));
+                        }
 
-                // found XmlSerializerAssemblyAttribute attribute, it should have all needed information to load the pre-generated serializer
-                if (assemblyAttribute.AssemblyName != null)
-                {
-                    serializerName = assemblyAttribute.AssemblyName;
-                    serializer = Assembly.Load(serializerName); // LoadWithPartialName just does this in .Net Core; changing the obsolete call.
-                }
-                else if (assemblyAttribute.CodeBase != null && assemblyAttribute.CodeBase.Length > 0)
-                {
-                    serializerName = assemblyAttribute.CodeBase;
-                    serializer = Assembly.LoadFrom(serializerName);
+                        return null;
+                    }
+
+                    if (!IsSerializerVersionMatch(serializer, type, defaultNamespace))
+                    {
+                        XmlSerializationEventSource.Log.XmlSerializerExpired(serializerName, type.FullName!);
+                        return null;
+                    }
                 }
                 else
                 {
-                    serializerName = type.Assembly.FullName;
-                    serializer = type.Assembly;
-                }
-                if (serializer == null)
-                {
-                    throw new FileNotFoundException(null, serializerName);
+                    System.Xml.Serialization.XmlSerializerAssemblyAttribute assemblyAttribute = (System.Xml.Serialization.XmlSerializerAssemblyAttribute)attrs[0];
+                    if (assemblyAttribute.AssemblyName != null && assemblyAttribute.CodeBase != null)
+                        throw new InvalidOperationException(SR.Format(SR.XmlPregenInvalidXmlSerializerAssemblyAttribute, "AssemblyName", "CodeBase"));
+
+                    // found XmlSerializerAssemblyAttribute attribute, it should have all needed information to load the pre-generated serializer
+                    if (assemblyAttribute.AssemblyName != null)
+                    {
+                        serializerName = assemblyAttribute.AssemblyName;
+                        serializer = Assembly.Load(serializerName); // LoadWithPartialName just does this in .Net Core; changing the obsolete call.
+                    }
+                    else if (assemblyAttribute.CodeBase != null && assemblyAttribute.CodeBase.Length > 0)
+                    {
+                        serializerName = assemblyAttribute.CodeBase;
+                        serializer = Assembly.LoadFrom(serializerName);
```

Comment:
Not sure what the intention is here, but this will probably not work correctly with custom ALCs.
`Assembly.LoadFrom` ALWAYS loads into the [default ALC](https://github.com/dotnet/runtime/blob/5fa6dd364982be4ffd83358adbf130d88049c72a/src/libraries/System.Private.CoreLib/src/System/Reflection/Assembly.cs#L367). I think this should call `AssemblyLoadContext.GetLoadContext(type.Assembly).LoadFromAssemblyPath` instead.

In short - it would be better to not use contextual reflection for this method at all.